### DISCUSSION TITLE
Build instructions for Android Studio

### DISCUSSION THREAD

**Discussion by bjornoleh** [NightscoutFoundation/xDrip#2041] [Q&A] [answered]
Hi, are there any build instructions generating an Xdrip APK with Android Studio available? I tried following what works for AAPS, but had build errors.

(Depending on build variant, one of them were: Unable to load class 'javax.xml.bind.JAXBException'.)

Hopefully it's not too different from building AAPS?
https://androidaps.readthedocs.io/en/latest/Installing-AndroidAPS/Building-APK.html#generate-signed-apk

I would just like to be able to build two instances of Xdrip with different app names and icon colours to easily differentiate an app for a left and right hand side transmitter. This is to be able to run overlapping transmitter sessions at day 9, and thus avoid poor data for looping.

I am aware of the build variants that are available on Google Drive. This will serve the purpose, but on a small screen (Jelly 2), I could not see the full app name under the icon. The only differentiation is then the location of the icon on the home screen.

https://github.com/Navid200/xDrip/wiki/Variants

https://drive.google.com/drive/folders/0B6mvYVNVC-fOWkxnVF80dlFabjQ?resourcekey=0-ebguuiPuB1wUI9Rp2zjMNg

**Comment by Lupum1001** [comment]
I got started within the past month or so, but my memory of how I solved it isn't perfect. I believe the problem with 'javax.xml.bind.JAXBException' was that the JDK that comes bundled with Android Studio is too new.

I think you need to install Java 8 from Oracle.com (also known as 1.8, I think?). The file I ended up downloading was called:
jdk-8u321-windows-x64.exe
https://www.oracle.com/java/technologies/downloads/#java8


I don't know what else I changed, but here are the settings that I'm currently using:
![image](https://user-images.githubusercontent.com/52087143/159690692-74793774-eaba-48e5-9496-bfa17044db24.png)
![image](https://user-images.githubusercontent.com/52087143/159690775-85caaf84-2cc5-49fe-94ef-66bff58506a8.png)
![image](https://user-images.githubusercontent.com/52087143/159691091-b3bb835d-7516-4852-8896-aadf917e1ec5.png)

Hope this helps!

**Reply by bjornoleh** [reply]
Thank you! I will give it a try. 

I forgot to mention my initial attempt was with AS on Mac. But I also have a working build environment on Windows, so I can test with both if needed.

**Comment by tolot27** [comment, marked as answer]
You can also merge #1876 or just cherry-pick 90ea672.

**Reply by bjornoleh** [reply]
Thank you @tolot27 ! I merged your Java11comp branch, and this allowed me to build on the Mac.

I would have tested cherry-picking the one commit too, but I'm a little fresh on Android studio (more familiar with Xcode and GitKraken for managing my git stuff).

**Comment by bjornoleh** [comment]
So I have managed to build Xdrip+ with AS now, but I have not been able to do what I intended to do when building. I hope it is ok to continue the conversation here, as it is included in the initial question.

My intention is to build my own build variants to have two instances of Xdrip that are easily distinguishable. This is in the context of running two Dexcom transmitters with overlapping sessions on the last day. I would like to have a second build with green icon (right hand side sensor) and possibly some distinguishing colouring inside of the app too. The original app with red icon can be for the left hand side.

I have tried adding productFlavour to build.gradle for both app and wear following these instructions:

https://developer.android.com/studio/build/build-variants#change-app-id

, and then adding new Image Assets with a green icon for both. 

My changes are visible in this branch:

https://github.com/bjornoleh/xDrip/commits/variants-bo

When building, bollowing the instructions for AAPS (Generate signed APK) I got the following error:

```
> Task :app:processGreenReleaseGoogleServices FAILED
Parsing json file: /Users/xxxxxx/AS/xDrip/app/google-services.json

Execution failed for task ':app:processGreenReleaseGoogleServices'.
> No matching client found for package name 'com.eveningoutpost.dexdrip.green'

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
```

**Reply by bjornoleh** [reply]
If someone would change the icon of one of the  build variants that are available, that would for course also be nice :-)

https://drive.google.com/drive/folders/0B6mvYVNVC-fOWkxnVF80dlFabjQ?resourcekey=0-ebguuiPuB1wUI9Rp2zjMNg

Here is a quick edit of the original icon:

![ic_launcher](https://user-images.githubusercontent.com/63544115/160104236-b34ace31-1ddb-440d-9500-a835fa8b5006.png)

**Reply by Navid200** [reply]
Never mind.  Your question is still about build errors.  It's perfectly fine to continue the conversation here.

**Reply by bjornoleh** [reply]
Thanks @Navid200


Do you happen know more about the build variants that are available from the Google Drive link? An additional version with a different icon would be great. But if I could find the repo from which the variants are buildt, I might be able to copy that for myself and just change the icons. 

I did this same thing pretty easily for XDrip4iOS, with two builds that are distinguishable by colour both by icon colour, name and colour inside the app. I was hoping it would be equally easy for  Xdrip+, but I am not very familiar with Android studio etc. 

With the Anubis transmitters becoming more readily available, I expect there to be some increase in the interest of what I am trying to achieve.

**Reply by Navid200** [reply]
My understanding is that the icon already has a different name.

**Reply by Navid200** [reply]
The colors inside the app can already be edited:
https://navid200.github.io/xDrip/docs/Legend.html

**Reply by Navid200** [reply]
I'm not 100% sure what you are trying to do.

If you intend to remove the 2-hour dead time every time you start a new sensor, you can already do it.
Let's say you start a new sensor once every 20 days.  Then, you only need a second app, in addition to your main app, for 2 hours once every 20 days.
During those 2 hours, the name on the icon will let you know which app you are opening.

**Reply by bjornoleh** [reply]
> 

Thanks, I didn't know about the colour editing options. Will look into that.

The dexcom G6 is often quite erratic for the first 8 - 24 hours, so it is not just the two hour warmup period. So the sensor will be started a day before the current one expire. We have done this for a while now, and have much less stress around sensor change than before. A new sensor can be noisy, more prone to compression lows, and sometimes require calibration. Now we can let it settle down before we start depending on it. 

This is for a kid, by the way. People who attempt full UAM looping also depend on having reliable CGM data all the time, as a jumpy sensor can be very challenging and slightly dangerous in that setting with an aggressively tuned APS.

The way we have done this on iOS, is that each transmitter is paired to its left (red) and right (green) instance of Xdrip4iOS. Then the new sensor can be started independently. It will upload to a spare NS site to let us easily check the state of the new sensor before we switch to this one. Then the old sensor expire, and we can close down that other app. This will be used to start a new sensor on the last day of the one that is currently running.

When done this way, we do not have to enter the transmitter ID every time, it is already set up. Also the NS urls are set, so its pretty easy to manage.

**Reply by bjornoleh** [reply]
With two apps on the phone, it is nice to immediately be able to tell them apart. On the very small Jelly 2 phone, I can not see the full name of the build variants, and the icons are the same.

**Reply by Navid200** [reply]
Thanks for the explanation.

I personally prefer if we add the capability to also change the color of the icon in the existing menu that allows changing other colors.
However, I don't know how practical this may be.  @jamorham

**Reply by Navid200** [reply]
Actually, I don't know if we should change the icon color.

Can you please show a screenshot of your phone where you cannot see the name of the variant?
There may be an alternative solution.

**Reply by bjornoleh** [reply]
Changing icon colour from the app would be great. Alternatively offering some alternatives in the build variants in Google Drive. 

Here is a photo of the tiny phone. 

![9BB6CC5C-BDC2-4795-92F7-AE1E59344583](https://user-images.githubusercontent.com/63544115/160133829-63ee88b9-9f68-4e59-9b0a-62910e33708a.jpeg)

**Reply by Navid200** [reply]
Perfect.  Thanks

@jamorham Can the variant identifier text precede xDrip in the title?

**Comment by Navid200** [comment]
There isn't an easy way to rename an installed app.  But, there are ways to do it.  I just did a search and got plenty of results.  One is to use a launcher, which lets you rename the app on the home screen.
You can do this to get what you need.

It seems to me if you add something in front of xDrip in the name, it should show on your screen.   For example, you can rename the variants to "L_xDrip" and "R_xDrip".

**Reply by bjornoleh** [reply]
Yes, that would do. Anything should be possible in Android Studio, but it proved a bit more difficult than anticipated based on my experience with Xcode/ iOS. And everyone won’t be able to do such things anyways.

**Reply by bjornoleh** [reply]
I might test a launcher, I saw those mentioned as well.

**Reply by Navid200** [reply]
I didn't mean in Android Studio.  I meant on your phone after having installed xDrip and the variant.
If a work-around like that works, you can use it.  And the developers can focus on fixing the bugs and really critical new features.

**Reply by Navid200** [reply]
Another approach could be to use widgets.  If you are willing to use some more real estate on the home screen, instead of having two icons, you can have two widgets.  
Again, you can control the background color of the widget using xDrip display settings.

**Reply by bjornoleh** [reply]
Thanks for the suggestion. The tiny phone can’t have two widgets side by side, that would have been pretty ideal. Here is the result after quickly setting some colours, This should work, but would probably be better on a bigger device:

![3E6400DA-4C1D-4F8D-92DE-0620F777AEFE](https://user-images.githubusercontent.com/63544115/160202803-3d4bee4d-9f52-445a-b4ad-cdfc84c102f3.jpeg)

The colour changes inside of the app works well too.

**Reply by Navid200** [reply]
If I think about the purpose of an icon, it is to represent some form of identity; the identity of an app.
Allowing the color of the icon to be changed kind of defeats the purpose of the icon to some extent.

I wish we could make the widget smaller.

I hope we can find a workaround.
Let's see if @jamorham has any suggestions.

**Reply by bjornoleh** [reply]
If you mean identity in terms of branding, then yes. But icons should also be easily distinguishable when two instances of an app can be installed side by side like here. If you look at my green version of the icon above, it is still pretty recognisable, and it carries some meaning with green vs red (right/ left or rather starboard/port). Not sure about a green drop of blood though, perhaps for diabetic lizards 😅 It would work for us anyways.

**Reply by bjornoleh** [reply]
For reference, this is what my versions of Xdrip4iOS looks like. It’s just my private modifications, I haven’t suggested it for the devs there yet:

![C54FB849-E0F1-423F-9A39-E00E9005117B](https://user-images.githubusercontent.com/63544115/160209139-dd9ed4e9-fe67-4147-b284-e4b7544a3b2c.jpeg)

The icon is originally white on blue background 

![6E290478-A770-40A2-9A6A-3681F1289D3E](https://user-images.githubusercontent.com/63544115/160209396-f5d56b66-5f9d-45ef-8897-35069ed96d44.png)

**Reply by bjornoleh** [reply]
I tested a tool that only change the icon and name of installed apps:

https://play.google.com/store/apps/details?id=com.alextern.shortcuthelper

This seems to work well. There is a workaround to avoid getting a supposed shortcut arrow-icon in the corner of the icon. This involved generating a widget instead of an icon (looks just like an icon). I tested the widget feature, and that works at least.

Icons can be added both as an icon and widget (perhaps both are technically widgets). These are displayed with different icon size, and the amount of visible text is slightly affected.

This is a screenshot of the tiny Jelly 2. Three icons/widgets are shown for each of the Xdrip+ variants (red = variant1, green = normal Xdrip+ install).

The smallest icons are the "widget" type.

I'd tweak the colouring of the Xdrip widget graphs if I were to use this, as these colours look too much like warning/OK colours when displayed together with CGM data. Currently this is just proof of concept. 

![image](https://user-images.githubusercontent.com/63544115/160371280-57daa07a-c467-4e25-a6fb-88974218034a.png)

The red/green colour coding inside of Xdrip seems fine:

![image](https://user-images.githubusercontent.com/63544115/160372420-3ffd128c-1400-42d8-a8ce-67b15b54138d.png)

![image](https://user-images.githubusercontent.com/63544115/160372518-fffdd3e5-75f5-4560-8014-244b7eed9623.png)

For now I am happy to have achieved what I set out to do. 

There are some talk about modified icons being reset by Android though, so I'll have to see if this works in the long run.

And I'd still be interested in some guidance on how to make build variants with different icons and names using Android studio. Alternatively have such variants included in the builds that are distributed on Google Drive. But this is low priority for me personally if my workaround proves reliable long term.

**Reply by bjornoleh** [reply]
Uploading the icons, in case anyone would like to reproduce this. I didn't spend too much time tweaking colours or anything, just consider this as an example.


![ic_launcher_red](https://user-images.githubusercontent.com/63544115/160391184-e66513d1-4066-405c-8ac9-9d7008439244.png)


<img width="212" alt="ic_launcher_red_on_white" src="https://user-images.githubusercontent.com/63544115/160391183-28d75bf6-6c7c-4908-ad44-b5c2c8e0e64a.png">

![ic_launcher_green](https://user-images.githubusercontent.com/63544115/160391530-bf7a5df4-c345-446f-b4d3-970995406aab.png)

<img width="212" alt="ic_launcher_green_on_white" src="https://user-images.githubusercontent.com/63544115/160391177-36df8d2b-03e2-48a5-b9ba-77e226cb2df7.png">

**Comment by jamorham** [comment]
The variants already have launcher titles like `xDrip+ va1` - I would suggest simply change the background color on the widget for one variant and tap the widget to enter the app. You don't need to use the launcher icon if you use widgets. I also recommend changing the `Lower Button Bar Flair` to the same color as the widget background (I use a purple color) so that when you are in the app there also cannot be any confusion as to which xDrip you are using.

**Reply by Navid200** [reply]
@jamorham
>The variants already have launcher titles

Yes, but on his phone, what comes after xDrip is not visible
![160133829-63ee88b9-9f68-4e59-9b0a-62910e33708a](https://user-images.githubusercontent.com/51497406/160413682-ef137d75-f450-4474-9f5e-50370f3868e8.jpeg)

**Reply by jamorham** [reply]
Well I think just use the regular and a single variant that can be differentiated as there would be no `v`, but I would suggest just using the widgets is very clear.
